jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,395 @@
+            if (cachedList.size() < cacheSize)
+                cachedList.add(entry);
+        }
+        return reusable;
+    }
+
+    public boolean isClosed()
+    {
+        return closed;
+    }
+
+    @Override
+    public void close()
+    {
+        closeLock.lock();
+        try
+        {
+            closed = true;
+
+            // clear the shared list and close its entries
+            for (Entry entry : sharedList)
+            {
+                if (remove(entry) && entry instanceof Closeable)
+                {
+                    try
+                    {
+                        ((Closeable)entry).close();
+                    }
+                    catch (IOException e)
+                    {
+                        LOGGER.warn("Error closing entry {}", entry, e);
+                    }
+                }
+            }
+
+            // clear the thread-local caches
+            allCaches.forEach(List::clear);
+            allCaches.clear();
+        }
+        finally
+        {
+            closeLock.unlock();
+        }
+    }
+
+    /**
+     * Remove a value from the pool.
+     *
+     * @param entry the value to remove
+     * @return true if the entry was removed, false otherwise
+     */
+    public boolean remove(Entry entry)
+    {
+        if (!entry.tryRemove() && !closed)
+        {
+            if (LOGGER.isDebugEnabled())
+                LOGGER.debug("Attempt to remove an object from the pool that is still in use: {}", entry);
+            return false;
+        }
+
+        boolean removed = sharedList.remove(entry);
+        if (!removed && !closed)
+        {
+            if (LOGGER.isDebugEnabled())
+                LOGGER.debug("Attempt to remove an object from the pool that does not exist: {}", entry);
+        }
+
+        return removed;
+    }
+
+    public int size()
+    {
+        return sharedList.size();
+    }
+
+    public Collection<Entry> values()
+    {
+        return Collections.unmodifiableCollection(sharedList);
+    }
+
+    @Override
+    public String toString()
+    {
+        return getClass().getSimpleName() + " size=" + sharedList.size() + " entries=" + sharedList;
+    }
+
+    public class Entry
+    {
+        private final AtomicBiInteger state = new AtomicBiInteger(); // hi: 0=open,1=closed lo: multiplexing counter
+        private final AtomicInteger usageCount = new AtomicInteger();
+        private final T pooled;
+
+        public Entry(T pooled)
+        {
+            this.pooled = pooled;
+        }
+
+        public T getPooled()
+        {
+            return pooled;
+        }
+
+        public boolean tryAcquire()
+        {
+            if (maxUsageCount >= 0)
+            {
+                int currentUsageCount = usageCount.getAndIncrement();
+                if (currentUsageCount < 0)
+                    currentUsageCount = Integer.MAX_VALUE;
+                if (currentUsageCount >= maxUsageCount)
+                    return false;
+            }
```

Comment:
I don't think usage count algorithm is correct.
1000 threads calling `pool.acquire()` will all hit the first `Entry` and increment the `usageCount` of this entry far beyond its multiplex value. The first thread that releases will see that the entry is over used although it's not.

Reply:
As discussed, I don't think that the algorithm is broken but it's tedious to reason about. So I've refactored it to store the three states in the `AtomicBiInteger` which should make it more maintainable.